plugins/resource_monitor_plugin/include/eosio/resource_monitor_plugin/file_space_handler.hpp

```diff
@@ -0,0 +1,143 @@
+               wlog( "Unable to get space info for ${path_name}: [code: ${ec}] ${message}. Ignire this failure.",
+                  ("path_name", fs.path_name.string())
+                  ("ec", ec.value())
+                  ("message", ec.message()));
+
+               continue;
+            }
+
+            if (info.available < fs.available_required) {
+               wlog("Space usage on ${path}'s file system exceeded configured threshold ${threshold}%, available: ${available}, Capacity: ${capacity}, available_required: ${available_required}", ("path", fs.path_name.string()) ("threshold", threshold) ("available", info.available) ("capacity", info.capacity) ("available_required", fs.available_required));
+
+               return true;
+            }
+         }
+
+         return false;
+      }
+
+      void add_file_system(const bfs::path& path_name) {
+         // Get detailed information of the path
+         struct stat statbuf;
+         auto status = space_provider.get_stat(path_name.string().c_str(), &statbuf);
+         EOS_ASSERT(status == 0, chain::plugin_config_exception,
+                    "Failed to run stat on ${path} with status ${status}", ("path", path_name.string())("status", status));
+
+         // If the file system containing the path is already
+         // in the filesystem list, do not add it again
+         for (auto& fs: filesystems) {
+            if (statbuf.st_dev == fs.st_dev) { // Two files belong to the same file system if their device IDs are the same.
+               ilog("${path_name}'s file system already monitored", ("path_name", path_name.string()));
+
+               return;
+            }
+         }
+
+         // For efficiency, precalculate threshold values to avoid calculating it
+         // everytime we check space usage. Since bfs::space returns
+         // available amount, we use minimum available amount as threshold. 
+         boost::system::error_code ec;
+         auto info = space_provider.get_space(path_name, ec);
+         EOS_ASSERT(!ec, chain::plugin_config_exception,
+            "Unable to get space info for ${path_name}: [code: ${ec}] ${message}",
+            ("path_name", path_name.string())
+            ("ec", ec.value())
+            ("message", ec.message()));
+
+         auto available_required = (100 - threshold) * (info.capacity / 100); // (100 - threshold) is the percentage of minimum number of available bytes the file system must maintain  
+
+         // Add to the list
+         filesystems.emplace_back(statbuf.st_dev, available_required, path_name);
+         
+         ilog("${path_name}'s file system monitored. available_required: ${available_required}, capacity: ${capacity}, threshold: ${threshold}", ("path_name", path_name.string()) ("available_required", available_required) ("capacity", info.capacity) ("threshold", threshold) );
+      }
+   
+   int space_monitor_loop() {
+      if ( is_threshold_exceeded() ) {
+         wlog("Shut down");
```

Comment:
Question: Should stopping be made an option of the plugin?
Also should it start logging a warning at some threshold before shutdown threshold?

Reply:
Will put out a warning at each percentage level 5% before shutdown.

Will make shutdown an option.